[DESK] Handle ".SoftwareSettings" section to fix VBoxVGA 7.0.12 driver installation

bobknier opened the pull request:
Added code to read the SoftwareSettings section of the VBoxVGA video driver file from the Virtual Box Additions installation.   I was unable to run ReactOS with Virtual Box Additions (version 7.0.12).

This PR aimed to fix JIRA issue: [CORE-19224](https://jira.reactos.org/browse/CORE-19224) but has proven to be wrong. In fact we do see a VirtualBox issue here which does also affect e.g. Windows 2003 Server SP2.
We will wait for https://www.virtualbox.org/ticket/21982 to be fixed instead.

HBelusca commented on dll/cpl/desk/classinst.c:
```diff
@@ -218,6 +218,33 @@ DisplayClassInstaller(
+    TCHAR SectionNameMod[MAX_PATH];
+    TCHAR *pSectionName;
+
+    StringCbCat(SectionNameMod, sizeof(SectionNameMod), SectionName);
```
Should be `StringCbCopy`

bobknier commented on dll/cpl/desk/classinst.c:
```diff
@@ -218,6 +218,33 @@ DisplayClassInstaller(
+    TCHAR *pSectionName;
+
+    StringCbCat(SectionNameMod, sizeof(SectionNameMod), SectionName);
+    while (SectionNameMod[i] < 'A' || SectionNameMod[i] > 'Z')
```
The loop is unecessary - I removed it and recommitted.    I also changed the string cat to a string copy.   Thanks for the input.

Attached is the INF file.   The section starts at line 75:

--- 
[VBoxVideo.SoftwareSettings]
AddReg = vboxvideo_SoftwareDeviceSettings

[vboxvideo_SoftwareDeviceSettings]
HKR,, InstalledDisplayDrivers, %REG_MULTI_SZ%, VBoxDisp
; Set 'Full' hardware acceleration up. On W2003 the level is by default None=5, leading to software mouse cursor
HKR,, Acceleration.Level, %REG_DWORD%, 0
HKR,, VgaCompatible, %REG_DWORD%, 0
----


[VBoxVideo.inf.zip](https://github.com/reactos/reactos/files/13438899/VBoxVideo.inf.zip)

ThFabba commented:
Could you explain in words what you're doing here please?
I understand that the code:
* takes the previous SectionName
* cuts it off at the first _
* appends `.SoftwareSettings`
* then installs that section

This seems suspicious, because I don't think underscores have special meaning in inf files. So we may be missing part of a more generic step here.
So it'd be good to get your understanding of what's missing that your change achieves.
Is the code somehow getting to `vboxvideo_SoftwareDeviceSettings` and you need to redirect it back to `VBoxVideo.SoftwareSettings`?

Thanks!

bobknier commented:
The section, VBoxVideo.SoftwareSettings is never referenced.    Also, the section VBoxVideo.OpenGLSoftwareSettings is not referenced.   I'm unable to find any documentation on the INF video file format to determine how these sections are parsed.    Maybe the [Manufacture] section should be parsed for the install-section-name (VBoxVideo)?

ThFabba commented:
> The section, VBoxVideo.SoftwareSettings is never referenced. Also, the section VBoxVideo.OpenGLSoftwareSettings is not referenced. I'm unable to find any documentation on the INF video file format to determine how these sections are parsed. Maybe the [Manufacture] section should be parsed for the install-section-name (VBoxVideo)?

Yes, exactly, `[Manufacturer]` references `VBoxVideo` and that should cause the inf parser to look for `VBoxVideo.SoftwareSettings` (and various other sections starting with `VBoxVideo.`). This then refers to `vboxvideo_SoftwareDeviceSettings` as its AddReg section, as Hermès mentioned.

It seems we have the code to add the `.SoftwareSettings` -- which should then lead our code to `vboxvideo_SoftwareDeviceSettings` if it's working correctly. So the fact that your change seems to be taking the section with an underscore and leading it _back to_ `.SoftwareSettings` is surprising. If you had any observations in terms of debug prints or other analysis, it would be very interesting to see that data.

bobknier commented:
Thank you for the comments.   I did modify the code (see the latest update) to use the manufacturer's section to obtain the correct section name.   I think this is more correct.   I initially used the previous technique (using the underscore) to put something together as a possible solution.   Does this make more sense?   This may not be the final solution, but should allow for integration to the latest version of Virtual Box Additions.

SergeGautherie reviewed (commented):
> `[Manufacturer]` references `VBoxVideo` and that should cause the inf parser to look for `VBoxVideo.SoftwareSettings` (and various other sections starting with `VBoxVideo.`)

Note:
Looking at
https://github.com/search?q=repo%3Areactos%2Freactos%20SoftwareSettings&type=code
1 .inf file (bochs) seems to match that design, 2 (pc98vid, xboxvmp) don't.
**To double-check**... (@binarymaster, @disean, what do you think of that?)

> debug prints or other analysis

Comparing registry content before/after install, on Windows/ReactOS, may help too.

SigmaTel71 commented on dll/cpl/desk/classinst.c:
```diff
@@ -4,7 +4,7 @@
  * PURPOSE:         Class installers
  *
- * PROGRAMMERS:     Herv� Poussineau (hpoussin@reactos.org)
+ * PROGRAMMERS:     Hervé Poussineau (hpoussin@reactos.org)
```
Check your file's codepage, this shouldn't have happened.

ThFabba replied:
This is a conversion _to_ UTF8, so is fine.

ThFabba commented:
Okay, I just ran through this in the debugger. Here's the data we get from setupapi:
```
kd> dx DeviceInfoData
DeviceInfoData                 : 0x12ea7c [Type: _SP_DEVINFO_DATA *]
    [+0x000] cbSize           : 0x1c [Type: unsigned long]
    [+0x004] ClassGuid        : {4D36E968-E325-11CE-BFC1-08002BE10318} [Type: _GUID]
    [+0x014] DevInst          : 0x1a [Type: unsigned long]
    [+0x018] Reserved         : 0x16ce50 [Type: unsigned long]
kd> dx DriverInfoData
DriverInfoData                 [Type: _SP_DRVINFO_DATA_V2_W]
    [+0x000] cbSize           : 0x61c [Type: unsigned long]
    [+0x004] DriverType       : 0x2 [Type: unsigned long]
    [+0x008] Reserved         : 0x170250 [Type: unsigned long]
    [+0x00c] Description      : "VirtualBox Graphics Adapter" [Type: wchar_t [256]]
    [+0x20c] MfgName          : "Oracle Corporation" [Type: wchar_t [256]]
    [+0x40c] ProviderName     : "Oracle Corporation" [Type: wchar_t [256]]
    [+0x60c] DriverDate       [Type: _FILETIME]
    [+0x614] DriverVersion    : 0x70000000c250c [Type: unsigned __int64]
kd> dx DriverInfoDetailData
DriverInfoDetailData                 [Type: _SP_DRVINFO_DETAIL_DATA_W]
    [+0x000] cbSize           : 0x622 [Type: unsigned long]
    [+0x004] InfDate          [Type: _FILETIME]
    [+0x00c] CompatIDsOffset  : 0x1 [Type: unsigned long]
    [+0x010] CompatIDsLength  : 0x0 [Type: unsigned long]
    [+0x014] Reserved         : 0x170250 [Type: unsigned long]
    [+0x018] SectionName      : "VBoxVideo_Install" [Type: wchar_t [256]]
    [+0x218] InfFileName      : "oem3.inf" [Type: wchar_t [260]]
    [+0x420] DrvDescription   : "VirtualBox Graphics Adapter" [Type: wchar_t [256]]
    [+0x620] HardwareID       : "" [Type: wchar_t [1]]
kd> dx SectionName
SectionName                 : "VBoxVideo_Install" [Type: wchar_t [260]]
```

So we're trying to install `VBoxVideo_Install.SoftwareSettings` instead of `VBoxVideo.SoftwareSettings` and that's why the change here makes it work.

Now the question is, is ReactOS's `SetupDiGetDriverInfoDetail` broken and returning the wrong section name, or should desk.cpl be using a different method to find the section name?

Looking at what Device Manager shows for "Inf Section" on Windows, it seems correct for this function to return the device section rather than the manufacturer section.

Searching for `.SoftwareSettings` in `C:\Windows\INF` yields nothing on my machine. But the documentation at https://learn.microsoft.com/en-us/windows-hardware/drivers/display/display-inf-file-sections gives an interesting hint: the section is supposed to be `DDInstall.SoftwareSettings`, which means `VBoxVideo_Install.SoftwareSettings` _is_ the correct section to be looking for.

Parsing a per-Manufacturer `.SoftwareSettings` section seems to be an undocumented feature. We'll have to establish whether Windows always does this, or only if the per-model section is absent. But given this... the current version of the fix doesn't seem far off the mark.

ThFabba commented:
Looking through my Win2003 INF directory, there are several `.SoftwareSettings` sections in there -- and they're all per-device, not per manufacturer. So arguably this is a bug in `VBoxVideo.inf`. But of course Windows seems to handle it, so we ought to as well...

bobknier commented:
ThFabba
    Good points.   Exactly what I found as well.   Without the change, ReactOS will throw an exception and not boot after applying the VirtualBoxAdditions.   Windows appears to handle it correctly, but the documentation is lacking.   I'm not sure how to continue.   It is more of how the ReactOS development should handle this type of situation.

HBelusca commented:
One venue of investigation is to replace our setupapi.dll with Windows 2003 one (+ one extra dll it depends upon), once ReactOS has been already installed and can reach its desktop. Then see whether the problem described here persists.

HBelusca reviewed (commented):
I've tried to observe what windows' own desk.cpl does:
basically it does the same calls as those we are doing originally in this function.
However, the code we have [between lines 53 to 70](https://github.com/reactos/reactos/pull/6026/files#diff-7143fdea7bf4192dad6356b3ace42cabe8e84d888f28cad3ace0ca4a66001ff2R53-R70) is executed later:
1. an extra `SetupDiGetDeviceInstallParams()` call +the flag-setting + the `SetupDiSetDeviceInstallParams()` call is done after having opened the InfFile and _before_ calling `SetupDiGetActualSectionToInstall()` on that mysterious `"DDInstall.SoftwareSettings"` section;
2. The `SetupDiInstallDevice()` call is made only at the very end, after having dealt with that `"DDInstall.SoftwareSettings"` section (included doing the OpenGLSoftwareSettings one, see around line 247 in our code).
3. Windows' desk.cpl doesn't appear to call `SetupDiGetDeviceRegistryProperty()` during this process to retrieve the service name.

While I don't think that points 2. and 3. fundamentally changes anything for what we are investigating, I'm wondering whether our code mainly at line 53+ (flag-setting + `SetupDiSetDeviceInstallParams()` call) _might_ modify what's going to be retrieved into `DriverInfoDetailData`, from which we derive that section name to which `.SoftwareSettings` gets appended.

HBelusca commented:
@ThFabba I'm wondering whether it could be possible to get similar setupapi traces as yours when trying to install that VBox driver inside a Windows 2003 VM :thinking:

ThFabba commented:
So uhm...
![image](https://github.com/reactos/reactos/assets/18138474/46a9f7b6-8e9e-4945-b6d9-ca96d1b8579b)
I think ReactOS is behaving correctly? These reg values do not get set on Windows.

ThFabba commented:
Changed the inf file to say `VBoxVideo_Install.SoftwareSettings` and:
![image](https://github.com/reactos/reactos/assets/18138474/6c62715a-bf36-424b-8fbc-68021154e763)

bobknier commented:
Ok.  Windows XP does add the additional registry settings.
The bigger problem is that after applying the VirtualBox additions, the ReactOS VM will no longer boot.   The display driver keeps the OS from boothing (seem image).


![ReactOSNoGo](https://github.com/reactos/reactos/assets/6306440/d85a008a-6681-4de8-8635-caaa7f06ded9)

HBelusca commented:
Which VBox version is this? Because iirc. there are problems with the additions with later ones. Maybe @hpoussin or @DarkFire01 could help you for this. However this seems to be offtopic vs. this PR.

bobknier commented:
Latest version (7.0..12r159484).   Using the latest guest additions (VBoxGuestAdditions_7.0.12.iso).


![image](https://github.com/reactos/reactos/assets/6306440/b88cc009-9d1d-4903-b3e8-e26c87a4cb1b)

JoachimHenze commented:
If this PR would be correct, that would mean that if we deleted the MS desk-cpl in MS 2k3sp2, the VBox installer would not longer complete there to install its VBoxVGA7.0.12 driver. I have not tested that yet, but I somehow can't believe that this is what we would see.
My gut-feeling is, that we maybe would touch the wrong module here with this fix.
Unless someone would prove the above test to actually break the installation on 2k3sp2, I will not approve this PR myself.

Nevertheless filing the PR was a good thing, because it gives us a better understanding/fingerpointing of what is currently preventing the proper installation.

bobknier commented:
Ok.   I downloaded and created another VM - Win2K3 SP2.   After successfully creating the VM, I installed the Virtual Box additions ISO and verfied the registry settings did not include the additional settings in the SoftwareSettings section of the driver - by looking in the registry.   This is the same behavior as ReactOS.   After rebooting the VM Windows 2003 SP2 did not boot in Virtual Box (black screen).

So here is what I found:   Windows XP works and does install the additional registry settings in the VBoxVGA settings of the registry.   Windows 2003 SP2 has the same behavior as ReactOS (does not install the additional registry settings and fails after a reboot).   Windows 7 uses a different driver (WDM).   It appears this is the same behavior as Windows 11.

At the end of the day, ReactOS is useless without the Virtual Box Additions.   The additions allow for integration with the host OS beyond the video driver (IMO).

JoachimHenze reviewed (changes requested):
see my comment above.

JoachimHenze commented:
If you delete the desk.cpl in Windows XPSP3 in the SFC-backup-folder, and then in system32, so that you can verify it is not longer present, and then try to install the driver in XPSP3, does it work then?

bobknier commented:
I reinstalled a Windows XP SP3 virtual machine, deleted both desk.cpl files (and verified that they were deleted).   I then applyed the VirtualBox additions (the drivers are bundled as a single application).   The result is that the video driver did install properly.   I verified that the registery settings were also present.   I then proceeded to find the desk.cpl files and they where found with the original dates (???).   

![image](https://github.com/reactos/reactos/assets/6306440/7a419dfc-deae-454a-923e-a96d49bcd6d4)

HBelusca commented:
> deleted both desk.cpl files [...] I then proceeded to find the desk.cpl files and they where found with the original dates (???).

When deleting the files both in the system32 and in system32\dllcache directories, when SFC is active the following behaviours can be observed:
1. If you delete the one in system32 first, the one in the dllcache will be used to restore it;
2. then if you delete the one in dllcache (and there is the file in system32 with the correct signature), the one in system32 will be used to restore the file in dllcache.

This is why you need to act very fast between the two deletions, during the small time lapse where the SFC detects the deletion in one directory, does its checks and then restore the file from the other directory.

The best way to do the deletion is to just stop the VM, mount its VHD in your main OS, do the file changes (then unmount the VHD), and restart the VM.

HBelusca commented:
> I confirmed they were deleted [...]

bobknier commented:
I tested the following senerios:

- Replaced the desk.cpl on the ReactOS VM with the Windows XP SP3 desk.cpl file.    I then applied the VirtualBox Additions.  Confirmed that the VirtualBox Graphics Adapter was installed.

![ReactOS](https://github.com/reactos/reactos/assets/6306440/12cd9683-d5b5-4ed1-bf19-d0828da72600)

- Replaced the desk.cpl on the Windows XP SP3 VM with the ReactOS desk.cpl file.   I then applied the VirtualBox Additions.   Confirmed that the VirtualBox Graphics Adapter was installed.

![image](https://github.com/reactos/reactos/assets/6306440/bfdd3fa7-2e27-44bc-8001-20fb421bafd8)

wjk commented:
This issue is still occurring today and needs to be fixed urgently. With all respect to the maintainers, I do not believe we should compromise the usability of our operating system in order to honor the patching process of a third-party project (which might not accept our patch at all to begin with). This is our bug, and it needs to be fixed.

SergeGautherie commented:
> Windows XP works and does install the additional registry settings in the VBoxVGA settings of the registry. Windows 2003 SP2 has the same behavior as ReactOS (does not install the additional registry settings and fails after a reboot)

This seems to be a "2K3 vs XP" issue.
Maybe ReactOS would want some kind of "XP shim" to support this case, if that is possible?

> I don't see any exceptions with windbg running Windows 2kSP3

Did you test on a _debug_ 2K3?

> I tested the following senerios:

This seems to show that XP/ROS desk.cpl is _not_ the key factor... (Could you do a XP/2K3 test, to confirm?)
Could the difference be in setupapi.dll, for example, as suggested before?

--

> This is our bug

It was stated that 2K3 has the same issue, so this does not look like a ROS bug to me.

JoachimHenze commented:
Further analysis is a job for the VBox guys: https://www.virtualbox.org/ticket/21982

JoachimHenze commented:
VirtualBox now fixed the bug with their most recent testbuilds https://www.virtualbox.org/wiki/Testbuilds
![VirtualBox_ReactOS_02_03_2024_18_55_56](https://github.com/reactos/reactos/assets/33393466/1795ae90-0f61-4547-9272-a23ce9c84bdd)

I will therefore close this PR. We do no longer need it to help the VBox-guys understand.

HBelusca commented:
Interesting analysis by the VBox and you guys in the VBox ticket!
I'm also surprised about this:
>  (Note that since 7.0.10, guest additions installer redirects XP32 guests towards the old-style INF -- but not XP64 or w2k3 guests) 

So this was the actual reason why it was mysteriously working on XP32, as observed by @bobknier here, but not elsewhere.

Good catch! :+1:

HBelusca commented:
@julenuri In the VBox ticket you said:
>  Post-ticket: Would it be ported back to the VBox GA 6 version? I know that is unmaintained or unsupported right now and maybe won't be sensible for the people that uses that old version. But if it would be possible, would be nice. 

Is the bug also in the VBox 6 version? (The reporter was mentioning VBox 7 here)
